### DISCUSSION TITLE
Time synchronization between Crazyflie and Anchors

### DISCUSSION THREAD

**Discussion by danteppc** [bitcraze/discussions#598] [Q&A]
Hi!

Let's say I have programmed this task with high priority in a Crazyflie and eight Loco nodes:

```
static uint32_t counter;
static void task (void *p) {
    while (1) {
        vTaskDelay(1);
	counter+=1;
    }
}

```
Obviously each device starts at different time so each one has a different counter. I need to bring this counter on all devices to sync within ±10ms error for a duration of 10 minutes.

I have a feeling that I should poll a system timer (like a monotonic one) instead of creating my own and then use some TWR protocol between pairs (Crazyflie being the center of star) to find the skew and calculate a new counter based on each device's UWB clock and the skew. Or maybe there is an easier way to achieve this like reset the clocks in DWM modules by broadcasting resync for a few seconds?  any thoughts?

Thank you.

**Comment by krichardsson** [comment]
The FreeRTOS function `xTaskGetTickCount()` will tell you the number of ticks the system has been up. On the Crazyflie one tick is 1 ms (you should use the `T2M()` macro to convert to ms anyway). As the clocks do not start at the same time they will return not be synchronized. 
The clocks will also drift, I'm not sure how much, but probably more than 10 ms in 10 minutes.

You will have to invent some way to synchronize your clocks, and most likely implement some sort of PLL in each node. I guess the most obvious communication path would be UWB. One option is to add a new LPP packet with synch data.

You can perhaps use the clock in the UWB packets as well, but it wraps quite often (not sure how often, but perhaps every second or so).

**Reply by danteppc** [reply]
Thank you for the info. I will look into taking advantage of UWB packets since they seem to be accurate enough for my application according to the DWM1000 datasheet:

> The module contains an on-board 38.4 MHz reference crystal. The crystal has been trimmed in production to reduce the initial frequency error to approximately 2 ppm

If I understand correctly, after 10 minutes the drift will be ±1.2ms=(10.60.1000).(±2/1000000).
And I assume `xTaskGetTickCount()` returns a less precise clock for generic purposes 🤔

**Reply by ataffanel** [reply]
The STM32 in the Crazyflie does not have a Crystal but a ceramic oscillator which is just-precise enough for USB to work, it has a drift of +/- 0.5%. This is the clock that is used by `xTaskGetTickCount()` so it will definitely not work for your purpose, the drift would be of +/-3000ms after 10min.

The general strategy for synchronizing multiple Crazyflie has usually been to send synchronization radio packets to them, ideally broadcast packets. This is something the UWB radio can definitely do. If your requirement is 'just' 10ms over 10min though, you need +/-16ppm and so just reading and keeping track of the DWM1000 clock should be good enough

Like @krichardsson says, the DWM1000 clock does wrap often, so you will need to have a mechanism to pool the timer value at regular interval to detect and update a 'wrap' counter in the Crazyflie. This should give you access to a 2ppm time reference.

If you need further synchronization between Crazyflies, you can start ranging between Crazyflies and it will allow you to synchronize clock between them by calculating and keeping updated an offset of your timer Vs. other Crazyflies timer.
